File: teste.py
import requests
from bs4 import BeautifulSoup
from pymongo import MongoClient
from flask import Flask, request, jsonify

app = Flask(__name__)

mongo_client = MongoClient('mongodb://localhost:27017/')
db = mongo_client['similarweb']
collection = db['website_data']

def scrape_similarweb(website_url):
    website_url = input()
    url = f'https://www.similarweb.com/website/{website_url}'
    response = requests.get(url)
    soup = BeautifulSoup(response.text, 'html.parser')

    title = soup.find('title').text
    category = soup.find('span', class_='websiteCategory').text
    global_rank = soup.find('span', class_='globalRank').text
    traffic_country = soup.find('div', class_='trafficCountry')
    top_countries = [country.text for country in traffic_country.find_all('div', class_='countryName')]


    data = {
        'website_url': website_url,
        'title': title,
        'category': category,
        'global_rank': global_rank,
        'top_countries': top_countries,
    }
    return data

@app.route('/salve_info', methods=['POST'])
def save_info():
    website_url = request.json.get('website_url')

    if not website_url:
        return jsonify({'error': 'URL do site ausente na solicitação'}), 400

    data = scrape_similarweb(website_url)

    collection.insert_one(data)

    return jsonify({'message': 'Scraping e armazenamento concluídos com sucesso!'})

@app.route('/get_info', methods=['POST'])
def get_info():
    website_url = request.json.get('website_url')

    if not website_url:
        return jsonify({'error': 'URL do site ausente na solicitação'}), 400

    data = collection.find_one({'website_url': website_url})

    if data:
        return jsonify(data)
    else:
        return jsonify({'error': 'Informações não encontradas no banco de dados'}), 404

if __name__ == '__main__':
    app.run(debug=True)

File: teste.rb
require 'sinatra'
require 'nokogiri'
require 'open-uri'
require 'mongo'
require 'selenium-webdriver'

configure do
  enable :sessions
  set :mongo_uri, 'mongodb://localhost:27017/similarweb_scraper'
end

before do
  content_type 'application/json'
end

post '/salve_info' do
  url = params[:url]
  halt 400, { error: 'URL is required' }.to_json unless url

  scraped_data = scrape_similarweb(url)
  save_to_mongo(url, scraped_data)

  { message: 'Data saved successfully' }.to_json
end

post '/get_info' do
  url = params[:url]
  halt 400, { error: 'URL is required' }.to_json unless url

  data = retrieve_from_mongo(url)
  halt 404, { error: 'Data not found' }.to_json unless data

  data.to_json
end

def scrape_similarweb(url)
  begin
    driver = Selenium::WebDriver.for :chrome
    driver.get("https://www.similarweb.com/website/#{url}")
    sleep(5)
    page = Nokogiri::HTML(driver.page_source)

    classification = page.xpath('//*[@id="overview"]/div/div/div/div[3]/div/div[1]/div/p').text.strip
    site = page.xpath('//*[@id="overview"]/div/div/div/div[1]/p[2]').text.strip
    category = page.xpath('//*[@id="overview"]/div/div/div/div[5]/div/dl/div[6]/dd/a').text.strip
    ranking_change = page.xpath('//*[@id="overview"]/div/div/div/div[3]/div/div[1]/div/span').text.strip
    average_visit_duration = page.xpath('//*[@id="overview"]/div/div/div/div[4]/div[2]/div[4]/p[2]').first.text.strip
    pages_per_visit = page.xpath('//*[@id="overview"]/div/div/div/div[4]/div[2]/div[3]/p[2]').last.text.strip
    bounce_rate = page.xpath('//*[@id="overview"]/div/div/div/div[4]/div[2]/div[2]/p[2]').text.strip
    
    driver.quit

    {
      classification: classification,
      site: site,
      category: category,
      ranking_change: ranking_change,
      average_visit_duration: average_visit_duration,
      pages_per_visit: pages_per_visit,
      bounce_rate: bounce_rate
    }

  rescue StandardError => e
    puts "Erro ao fazer scraping: #{e.message}"
    {
      classification: '',
      site: '',
      category: '',
      ranking_change:'',
      average_visit_duration: '',
      pages_per_visit: '',
      bounce_rate: ''
    }
  end
end

def save_to_mongo(url, data)

  client = Mongo::Client.new(settings.mongo_uri)
  collection = client[:websites]

  existing_data = collection.find(url: url).first
  if existing_data
    collection.update_one({ _id: existing_data['_id'] }, '$set' => data)
  else
    collection.insert_one(data.merge(url: url))
  end
end

def retrieve_from_mongo(url)
  client = Mongo::Client.new(settings.mongo_uri)
  collection = client[:websites]

  collection.find(url: url).first
end
